File: skills/tools/eas-planning-writer/scripts/test_planning_writer.py
#!/usr/bin/env python3
"""
eas-planning-writer 技能测试脚本
验证所有脚本是否正常工作并与技能定义保持一致
"""

import os
import subprocess
import sys
from pathlib import Path
from typing import List, Tuple, Optional

def run_command(cmd: str, cwd: Optional[str] = None) -> Tuple[int, str, str]:
    """运行命令并返回输出"""
    try:
        result = subprocess.run(
            cmd, 
            shell=True, 
            capture_output=True, 
            text=True, 
            cwd=cwd
        )
        return result.returncode, result.stdout, result.stderr
    except Exception as e:
        return -1, "", str(e)

def test_script_execution(script_path: str, test_args: Optional[List[str]] = None) -> bool:
    """测试脚本是否能正常执行"""
    if not os.path.exists(script_path):
        print(f"Script does not exist: {script_path}")
        return False
    
    print(f"Testing script execution: {script_path}")
    
    cmd = f"python {script_path}"
    if test_args:
        cmd += " " + " ".join(test_args)
    
    return_code, stdout, stderr = run_command(cmd)
    
    if return_code == 0:
        print(f"SUCCESS: Script executed successfully: {script_path}")
        return True
    else:
        print(f"FAILED: Script execution failed: {script_path}")
        print(f"   Return code: {return_code}")
        print(f"   Stdout: {stdout}")
        print(f"   Stderr: {stderr}")
        return False

def test_python_scripts():
    """测试 Python 脚本"""
    print("\nTesting Python scripts")
    print("="*50)
    
    scripts_dir = Path(__file__).parent
    results = {}
    
    # 测试 session-catchup.py
    script_path = scripts_dir / "session-catchup.py"
    results["session-catchup.py"] = test_script_execution(str(script_path), ["--help"]) if script_path.exists() else False
    
    # 测试新添加的脚本
    init_script = scripts_dir / "init_planning_session.py"
    results["init_planning_session.py"] = test_script_execution(str(init_script), ["--help"]) if init_script.exists() else False
    
    check_script = scripts_dir / "check_complete.py"
    results["check_complete.py"] = test_script_execution(str(check_script), ["--help"]) if check_script.exists() else False
    
    return results

def test_shell_scripts():
    """测试 Shell/Bash 脚本"""
    print("\nTesting Shell/Bash scripts")
    print("="*50)
    
    scripts_dir = Path(__file__).parent
    results = {}
    
    # 测试 check-complete.sh
    script_path = scripts_dir / "check-complete.sh"
    if script_path.exists():
        print(f"Testing script execution: {script_path}")
        return_code, stdout, stderr = run_command(f"bash {script_path}")
        if return_code == 0 or return_code == 1:  # 1 通常是正常情况（未找到task_plan.md）
            print(f"SUCCESS: Script executed successfully: {script_path}")
            results["check-complete.sh"] = True
        else:
            print(f"FAILED: Script execution failed: {script_path}")
            print(f"   Return code: {return_code}, Error: {stderr}")
            results["check-complete.sh"] = False
    else:
        print(f"Script does not exist: {script_path}")
        results["check-complete.sh"] = False
    
    # 测试 init-session.sh
    script_path = scripts_dir / "init-session.sh"
    if script_path.exists():
        print(f"Testing script execution: {script_path}")
        return_code, stdout, stderr = run_command(f"bash {script_path}")
        if return_code == 0 or "Planning files initialized" in stdout:
            print(f"SUCCESS: Script executed successfully: {script_path}")
            results["init-session.sh"] = True
        else:
            print(f"FAILED: Script execution failed: {script_path}")
            print(f"   Return code: {return_code}, Error: {stderr}")
            results["init-session.sh"] = False
    else:
        print(f"Script does not exist: {script_path}")
        results["init-session.sh"] = False
    
    return results

def test_functionality():
    """测试功能完整性"""
    print("\nTesting functionality")
    print("="*50)
    
    # 创建临时目录进行测试
    import tempfile
    import shutil
    
    with tempfile.TemporaryDirectory() as temp_dir:
        print(f"Using temp directory: {temp_dir}")
        
        # 测试创建规划文件
        init_script = Path(__file__).parent / "init_planning_session.py"
        if init_script.exists():
            print("Testing planning file creation...")
            
            # 创建 docs/task 目录
            task_dir = Path(temp_dir) / "docs" / "task"
            task_dir.mkdir(parents=True, exist_ok=True)
            
            # 运行初始化脚本
            cmd = f"python {init_script} --output {task_dir}"
            return_code, stdout, stderr = run_command(cmd, cwd=temp_dir)
            
            if return_code == 0:
                # 检查文件是否创建成功
                task_plan = task_dir / "task_plan.md"
                findings = task_dir / "findings.md"  
                progress = task_dir / "progress.md"
                
                success_count = 0
                for file_path in [task_plan, findings, progress]:
                    if file_path.exists():
                        print(f"SUCCESS: File created: {file_path.name}")
                        success_count += 1
                    else:
                        print(f"FAILED: File creation failed: {file_path.name}")
                
                if success_count == 3:
                    print("SUCCESS: Planning file creation functionality intact")
                    return True
                else:
                    print("FAILED: Planning file creation functionality is incomplete")
                    return False
            else:
                print(f"FAILED: Planning file creation script execution failed: {stderr}")
                return False
        else:
            print("Initialization script does not exist, skipping functionality test")
            return True

def main():
    """主测试函数"""
    print("EAS-PLANNING-WRITER Skill Validation Test")
    print("="*60)
    
    # 测试 Python 脚本
    python_results = test_python_scripts()
    
    # 测试 Shell 脚本
    shell_results = test_shell_scripts()
    
    # 测试功能完整性
    functionality_result = test_functionality()
    
    # 汇总结果
    print("\nTest Results Summary")
    print("="*60)
    
    all_results = {**python_results, **shell_results}
    
    for script, success in all_results.items():
        status = "SUCCESS" if success else "FAILED"
        print(f"[{status}] {script}: {'PASS' if success else 'FAIL'}")
    
    print(f"\nFunctionality test: {'SUCCESS' if functionality_result else 'FAILED'}")
    
    # 计算成功率
    total_tests = len(all_results) + (1 if functionality_result is not None else 0)
    passed_tests = sum(1 for result in all_results.values() if result) + (1 if functionality_result else 0)
    
    print(f"\nOverall Pass Rate: {passed_tests}/{total_tests} ({passed_tests/total_tests*100:.1f}%)")
    
    if all(result for result in all_results.values()) and functionality_result:
        print("\nAll tests passed! eas-planning-writer skill validation successful.")
        return 0
    else:
        print(f"\nSome tests failed. Please check if scripts align with skill definition.")
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: skills/buildin/eas-planning-writer/scripts/test_planning_writer.ts
#!/usr/bin/env tsx
/**
 * eas-planning-writer 技能测试脚本
 * 验证所有 .ts 脚本是否正常工作并与技能定义保持一致
 */

import { spawnSync } from 'child_process';
import { promises as fs } from 'fs';
import * as os from 'os';
import * as path from 'path';
import { fileURLToPath } from 'url';

interface TestResult {
  name: string;
  pass: boolean;
  stdout: string;
  stderr: string;
  exitCode: number;
}

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

function runScript(scriptRelPath: string, args: string[] = []): { code: number; stdout: string; stderr: string } {
  const scriptPath = path.join(__dirname, scriptRelPath);
  const result = spawnSync('npx', ['tsx', scriptPath, ...args], {
    encoding: 'utf-8',
    shell: process.platform === 'win32',
  });
  return {
    code: result.status ?? -1,
    stdout: result.stdout ?? '',
    stderr: result.stderr ?? '',
  };
}

function testInitScript(): TestResult {
  // 烟雾测试：调用 --help 应退出 0
  const r = runScript('init-planning-session.ts', ['--help']);
  return {
    name: 'init-planning-session.ts --help',
    pass: r.code === 0,
    stdout: r.stdout,
    stderr: r.stderr,
    exitCode: r.code,
  };
}

async function testInitFunctionality(): Promise<TestResult> {
  // 创建临时目录，运行 init 脚本，检查三件套是否生成
  const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), 'eas-planning-test-'));
  const taskDir = path.join(tempDir, 'task');
  try {
    const r = runScript('init-planning-session.ts', ['--output', taskDir]);
    if (r.code !== 0) {
      return {
        name: 'init-planning-session.ts 落盘三件套',
        pass: false,
        stdout: r.stdout,
        stderr: r.stderr,
        exitCode: r.code,
      };
    }
    const expected = ['task_plan.md', 'findings.md', 'progress.md'];
    const checks = await Promise.all(
      expected.map(async (f) => {
        try {
          await fs.access(path.join(taskDir, f));
          return true;
        } catch {
          return false;
        }
      }),
    );
    const allOk = checks.every(Boolean);
    return {
      name: 'init-planning-session.ts 落盘三件套',
      pass: allOk,
      stdout: r.stdout,
      stderr: r.stderr,
      exitCode: 0,
    };
  } finally {
    await fs.rm(tempDir, { recursive: true, force: true });
  }
}

function testCheckCompleteScript(): TestResult {
  // 无参数：扫描 cwd 的 task_plan.md
  const r = runScript('check-complete.ts');
  return {
    name: 'check-complete.ts (无参数)',
    pass: r.code === 0 || r.code === 1, // exit 0/1 都是正常
    stdout: r.stdout,
    stderr: r.stderr,
    exitCode: r.code,
  };
}

async function testCheckCompleteWithFile(): Promise<TestResult> {
  // 准备一个 task_plan.md，验证报告输出
  const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), 'eas-planning-test-'));
  const planFile = path.join(tempDir, 'task_plan.md');
  const content = `# Test

## 阶段 (Phases)

### Phase 1: One
- [ ] item
- **Status:** complete

### Phase 2: Two
- [ ] item
- **Status:** in_progress
`;
  try {
    await fs.writeFile(planFile, content, 'utf-8');
    const r = runScript('check-complete.ts', [planFile]);
    const ok = r.code === 0 && /1\/2 phases complete/.test(r.stdout);
    return {
      name: 'check-complete.ts 报告阶段完成度',
      pass: ok,
      stdout: r.stdout,
      stderr: r.stderr,
      exitCode: r.code,
    };
  } finally {
    await fs.rm(tempDir, { recursive: true, force: true });
  }
}

function testSessionCatchupScript(): TestResult {
  // 无项目路径参数：不崩溃即可
  const r = runScript('session-catchup.ts', [os.tmpdir()]);
  return {
    name: 'session-catchup.ts (无项目历史应静默退出)',
    pass: r.code === 0,
    stdout: r.stdout,
    stderr: r.stderr,
    exitCode: r.code,
  };
}

function printResult(r: TestResult): void {
  const tag = r.pass ? '[PASS]' : '[FAIL]';
  console.log(`${tag} ${r.name}`);
  if (!r.pass) {
    if (r.stderr) console.log(`  stderr: ${r.stderr.trim()}`);
    if (r.stdout) console.log(`  stdout: ${r.stdout.trim()}`);
    console.log(`  exitCode: ${r.exitCode}`);
  }
}

async function main(): Promise<void> {
  console.log('EAS-PLANNING-WRITER Skill Validation Test');
  console.log('='.repeat(60));

  const results: TestResult[] = [];
  results.push(testInitScript());
  results.push(await testInitFunctionality());
  results.push(testCheckCompleteScript());
  results.push(await testCheckCompleteWithFile());
  results.push(testSessionCatchupScript());

  console.log('\nTest Results');
  console.log('='.repeat(60));
  for (const r of results) printResult(r);

  const passed = results.filter((r) => r.pass).length;
  const total = results.length;
  console.log(`\nOverall: ${passed}/${total} (${((passed / total) * 100).toFixed(1)}%)`);

  if (passed === total) {
    console.log('\nAll tests passed! eas-planning-writer skill validation successful.');
    process.exit(0);
  } else {
    console.log(`\n${total - passed} test(s) failed. Please check the .ts scripts.`);
    process.exit(1);
  }
}

// 只有在直接运行此脚本时才执行主函数（ESM 入口检测）
const isDirectRun = (() => {
  try {
    return process.argv.some((arg) => arg === __filename || arg.endsWith(__filename.replace(/\\/g, '/')));
  } catch {
    return false;
  }
})();

if (isDirectRun) {
  main().catch((error: unknown) => {
    const message = error instanceof Error ? error.message : String(error);
    console.error('Test runner error:', message);
    process.exit(1);
  });
}

export { runScript, testInitScript, testCheckCompleteScript, testSessionCatchupScript };
